File: Python/OCR.py
## 알고리즘도 맞고, 여기서 돌려봤을 때는 아무 문제가 없는 코드인데, 채점 서버에서 돌리면 입출력에서부터 에러가 난다.
## 거지같은 알고스팟 ㅅㅄㅄㅄㅅㅄㅄㅄㅄㅄㅄㅄㅄ
## 결국 C++로 옮겨서 해결, 나중에 시간나면 고쳐서 다시 해봐야겠다.

import sys
import math
from functools import reduce

memo = []
sys.setrecursionlimit = 10**8

def get_prob(s_index, prev_word):
    """
    주어진 문장에서의 시작 index (단어의 index)와 그 전의 word가 주어졌을 때 최대 확률을 계산
    """
    result = 0
    # 시작일 경우
    if s_index >= sen_len:
        return 1.0
    if memo[s_index][prev_word] != -1:
        return memo[s_index][prev_word]
    cur_word_ind = words.index(sen[s_index])
    if s_index == 0:
        for i in range(no_of_word):
            cur_prob = first_word_prob[i] * wrong_word_prob[i][cur_word_ind]
            if cur_prob != -math.inf:
                result = max(result, cur_prob * get_prob(s_index + 1, i))
    else:
        for i in range(no_of_word):
            cur_prob = next_word_prob[prev_word][i] * wrong_word_prob[i][cur_word_ind]
            if cur_prob !=-math.inf:
                result = max(result, cur_prob * get_prob(s_index + 1, i))
    memo[s_index][prev_word] = result
    return result

def reconstruct(s_index, prev_word, answer):
    if s_index >= len(sen):
        return
    cur_word_ind = words.index(sen[s_index])
    if s_index == 0:
        for i in range(no_of_word):
            cur_prob = first_word_prob[i] * wrong_word_prob[i][cur_word_ind]
            result = cur_prob * get_prob(s_index + 1, i)
            if result == get_prob(s_index, 0):
                answer.append(words[i])
                reconstruct(s_index + 1, i, answer)
                break
    else:
        for i in range(no_of_word):
            cur_prob = next_word_prob[prev_word][i] * wrong_word_prob[i][cur_word_ind]
            result = cur_prob * get_prob(s_index + 1, i)
            if result == get_prob(s_index, prev_word):
                answer.append(words[i])
                reconstruct(s_index + 1, i, answer)
                break


if __name__ == "__main__":
    ##
    # no_of_word, no_of_sentence = map(int, input().strip().split())
    # words = input().strip().split()
    # first_word_prob = [float(x) for x in sys.readline().strip().split()]
    # next_word_prob = []
    # wrong_word_prob = []
    # sen_len = 0
    # sen = []
    ##

    no_of_word, no_of_sentence = [int(x) for x in sys.stdin.readline().strip().split()]
    words = sys.stdin.readline().strip().split()
    first_word_prob = [float(x) for x in sys.stdin.readline().strip().split()]
    next_word_prob = [0] * no_of_word 
    wrong_word_prob = [0] * no_of_word
    sen_len = 0
    sen = []


    for i in range(no_of_word):
        next_word_prob[i] = [float(x) for x in sys.stdin.readline().strip().split()]
    for i in range(no_of_word):
        wrong_word_prob[i] = [float(x) for x in sys.stdin.readline().strip().split()]

    for _ in range(no_of_sentence):
        tmp = sys.stdin.readline().strip().split()
        sen_len = int(tmp[0])
        sen = tmp[1:]
        memo = [[-1] * no_of_word for _ in range(sen_len)]
        answer = [] 
        reconstruct(0, 0, answer)
        # print(reduce(lambda x, y: x + " "  + y, answer))

File: C++/OCR.cpp
// 그렇게 어려운 문제는 아닌데, CPP의 사용법 때문에 힘들었던 문제.
// Underflow를 막기 위해 log를 취하고, 곱하기를 덧셈으로 바꾸는 방식은 잘 기억해두자.
// 파이썬으로도 어떻게 되는거 같긴 한데, 알 수 없는 이유때문에 Nonzero return code가 뜬다.
// 조건부 확률이 잘 기억나진 않았지만, 그냥 본능대로 식을 세워 푸니 맞았다.

#include <iostream>
#include <vector>
#include <algorithm>
#include <cstring>
#include <bits/stdc++.h>
using namespace std;


#define INF 999999.0
int no_of_word, no_of_sentence, sen_len;
string ind_to_word[501], sen[102];
double first_word_prob[501], next_word_prob[501][501], wrong_word_prob[501][501], memo[501][101];
double get_prob(int, int);
void reconstruct(int, int, string&);

int main() {
    int i,j,k;
    scanf("%d %d", &no_of_word, &no_of_sentence);
    for(i=0; i<no_of_word; i++)
        cin >> ind_to_word[i];
    for(i=0; i<no_of_word; i++){
        cin >> first_word_prob[i];
        first_word_prob[i] = log(first_word_prob[i]);
    }
    for(i=0; i<no_of_word; i++){
        for(j=0; j<no_of_word; j++){
            cin >> next_word_prob[i][j];
            next_word_prob[i][j] = log(next_word_prob[i][j]);
        }
    }
    for(i=0; i<no_of_word; i++){
        for(j=0; j<no_of_word; j++){
            cin >> wrong_word_prob[i][j];
            wrong_word_prob[i][j] = log(wrong_word_prob[i][j]);
        }
    }
    for(i=0; i<no_of_sentence; i++){
        cin >> sen_len;
        for(j=0; j<sen_len; j++){
            cin >> sen[j];
        }
        // fill(begin(memo), end(memo), 1.0);
        for(k=0; k<501; k++){
            for(int l=0; l<101; l++){
                memo[k][l] = 1;
            }
        }
        // memset(memo, 1, sizeof(memo));
        string answer = "";
        reconstruct(0, 0, answer);
        std::cout << answer << std::endl;
    }
    return 0;
}

double get_prob(int s_index, int prev_word_ind){
    int i;
    if(s_index >= sen_len){
        return 1.0;
    }
    double& result = memo[prev_word_ind][s_index];
    if(result != 1)
        return result;
    result = -INF;
    int cur_word_ind = distance(ind_to_word, find(ind_to_word, ind_to_word + 501, sen[s_index]));
    if(s_index == 0){
        for(i=0; i<no_of_word; i++){
            double cur_prob = first_word_prob[i] + wrong_word_prob[i][cur_word_ind];
            result = max(result, cur_prob + get_prob(s_index + 1, i));
        }
    }
    else{
        for(i=0; i<no_of_word; i++){
            double cur_prob = next_word_prob[prev_word_ind][i] + wrong_word_prob[i][cur_word_ind];
            result = max(result, cur_prob + get_prob(s_index + 1, i));
        }
    }
    return result;
}

void reconstruct(int s_index, int prev_word_ind, string &answer){
    int i;
    if(s_index >= sen_len)
        return;
    int cur_word_ind = distance(ind_to_word, find(ind_to_word, ind_to_word + 501, sen[s_index]));
    if(s_index == 0){
        for(i=0; i<no_of_word; i++){
            double cur_prob = first_word_prob[i] + wrong_word_prob[i][cur_word_ind];
            double result = cur_prob + get_prob(s_index + 1, i);
            if(result == get_prob(s_index, 0)){
                answer += ind_to_word[i];
                answer += " ";
                reconstruct(s_index + 1, i, answer);
                break;
            }
        }
    }
    else{
        for(i=0; i<no_of_word; i++){
            double cur_prob = next_word_prob[prev_word_ind][i] + wrong_word_prob[i][cur_word_ind];
            double result = cur_prob + get_prob(s_index + 1, i);
            if(result == get_prob(s_index, prev_word_ind)){
                answer += ind_to_word[i];
                answer += " ";
                reconstruct(s_index + 1, i, answer);
                break;
            }
        }
    }
}